crypto/schnorr-adaptor-signature/sign_utils.go

```diff
@@ -9,168 +9,248 @@ import (
+	if (T.X.IsZero() && T.Y.IsZero()) || T.Z.IsZero() {
+		return fmt.Errorf("T point is at infinity")
 	}
-	// ensure R' is same as the expected R' = s'*G - e*P by checking both x and y coordinates
-	if !expRHat.X.Equals(&RHat.X) || !expRHat.Y.Equals(&RHat.Y) {
-		return fmt.Errorf("expected R' = s'*G - e*P is different from the actual R'")
+	T.ToAffine()
+
+	// Step 7: Verify T matches encryption key
+	T.ToAffine()
+	if !T.X.Equals(&t.X) || !T.Y.Equals(&t.Y) {
+		return fmt.Errorf("extracted encryption key does not match")
 	}
 
 	return nil
 }
 
-// intoPointWithEvenY converts the given Jacobian point to an affine
-// point with even y value, and returns a bool value on whether the
-// negation is performed.
-// The bool value will be used for decrypting an adaptor signature
-// to a Schnorr signature.
-func intoPointWithEvenY(point *btcec.JacobianPoint) (*btcec.JacobianPoint, bool) {
-	affinePoint := point
-	affinePoint.ToAffine()
+// decrypt decrypts a pre-signature using a decryption key.
+// This implements the Decrypt algorithm as defined in the spec.
+func decrypt(psig *AdaptorSignature, dk *btcec.ModNScalar) (*schnorr.Signature, error) {
+	// Get s value from adaptor signature
+	s0 := psig.s
 
-	needNegation := affinePoint.Y.IsOdd()
+	// Get decryption key value
+	t := *dk
 
-	if needNegation {
-		affinePoint = negatePoint(affinePoint)
+	// Check if s0 or t is zero or exceeds curve order
+	if s0.IsZero() || t.IsZero() {
+		return nil, fmt.Errorf("s0 or decryption key is zero")
 	}
 
-	return affinePoint, needNegation
+	// Get R point from adaptor signature
+	R0 := psig.R0
+	R0.ToAffine()
+
+	// Compute final s value based on R0.Y being even/odd
+	// This matches the Python reference implementation:
+	// if sig[0] == 2 (even y): s = (s0 + t) % n
+	// if sig[0] == 3 (odd y): s = (s0 - t) % n
+	var s btcec.ModNScalar
+	if !R0.Y.IsOdd() {
+		// R0.Y is even (sig[0] == 2)
+		// s = (s0 + t) % n
+		s.Set(&s0)
+		s.Add(&t)
+	} else {
+		// R0.Y is odd (sig[0] == 3)
+		// s = (s0 - t) % n
+		s.Set(&s0)
+		tNeg := t
+		tNeg.Negate()
+		s.Add(&tNeg)
+	}
+
+	// Create Schnorr signature from R0.X and s
+	return schnorr.NewSignature(&R0.X, &s), nil
+}
+
+// extract extracts the decryption key from a pre-signature and its decrypted signature.
+// This implements the Extract algorithm as defined in the spec.
+func extract(psig *AdaptorSignature, sig *schnorr.Signature) (*btcec.ModNScalar, error) {
+	// Get s0 value from adaptor signature
+	s0 := psig.s
+
+	// Get s value from schnorr signature
+	sigBytes := sig.Serialize()
+	sBytes := sigBytes[32:]
+	var s btcec.ModNScalar
+	if overflow := s.SetByteSlice(sBytes); overflow {
+		return nil, fmt.Errorf("s value in signature is invalid")
+	}
+
+	// Check if s0 or s is zero or exceeds curve order
+	if s0.IsZero() || s.IsZero() {
+		return nil, fmt.Errorf("s values must be non-zero")
+	}
+
+	// Get R0 point from adaptor signature to check its Y parity
+	R0 := psig.R0
+	R0.ToAffine()
```

Comment:
we have this `R0.ToAffine()` methods scattered around... 

Can we define an invariant for `AdaptorSignature` struct that it always contains point in affine cooridnates and check it during struct construction ?

Reply:
Hmm actually this invariant is already the case https://github.com/babylonlabs-io/babylon/blob/51d86025be02ac7d4821f2abba4e0e84bb5abddc/crypto/schnorr-adaptor-signature/sign_utils.go#L44-L45. We can remove L224, and add some docstring on this invariant